### PR TITLE
Calendrical calculations move for islamic, hebrew, and chinese_based, and polish

### PR DESCRIPTION
**Pull request by Manishearth**
fixes https://github.com/unicode-org/icu4x/issues/3975

### CODE REVIEW

File: utils/calendrical_calculations/src/islamic.rs [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+            + libm::floor(((year as f64 - 1.0) * 12.0 + month as f64 - 0.5) * MEAN_SYNODIC_MONTH)
+                as i64,
+    );
+    let first_day_of_month = saudi_new_month_on_or_before(midmonth).to_i64_date();
+
+    RataDie::new(first_day_of_month + day as i64 - 1)
+}
+
+// Lisp code reference:https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L2076
+pub fn fixed_from_islamic_civil(year: i32, month: u8, day: u8) -> RataDie {
```
[the remark this answers is not in the document]
Comment (Manishearth) [reply utils/calendrical_calculations/src/islamic.rs:109]: Trying to match the book here mostly. This is something we should try later IMO. There's a lot of cleanups we can perform here, some of which I plan to do after this refactoring is over, but I don't want to block the refactoring on this.

**Comment by robertbastian** [comment]
We have a Polish calendar now? 😁

**Comment by Manishearth** [comment]
yeah it's a present for @zbraniecki

File: components/calendar/src/hebrew.rs [hunk 2]
```diff
@@ -309,64 +279,22 @@ impl Hebrew {
...
-            day: civil_date.0.day,
-        }
+        BookHebrew::from_civil_date(civil_date.0.year, civil_date.0.month, civil_date.0.day)
     }
 
-    // ADAR = 12, ADARII = 13
     fn last_month_of_civil_hebrew_year(civil_year: i32) -> u8 {
         if Self::is_leap_year(civil_year) {
-            ADARII
+            13 // there are 13 months in a leap year
```
Comment (echeran) [inline components/calendar/src/hebrew.rs:295]: Out of curiosity, why are the `u8` vlaues used directly here, but the similar function for the last month of the book Hebrew year use the month name constants?

File: utils/calendrical_calculations/src/islamic.rs [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+    latitude: 30.1,
+    longitude: 31.3,
+    elevation: 200.0,
+    zone: (1_f64 / 12_f64),
+};
+
+// "Fixed" is a day count representation of calendars staring from Jan 1st of year 1 of the Georgian Calendar.
+// The fixed date algorithms are from
+// Dershowitz, Nachum, and Edward M. Reingold. _Calendrical calculations_. Cambridge University Press, 2008.
+/// Lisp code reference: <https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L6904>
```
Comment (echeran) [inline utils/calendrical_calculations/src/islamic.rs:21]: question: Which comments should be Rustdoc vs. "regular" non-doc?

The pattern that I observe in this module is that the "Lisp code reference: ..." comments are Rustdoc, but if there is an explanatory comment, it doesn't make it into the Rust docs. But changing the explanatory comments to be included in Rustdocs seems preferable.

File: utils/calendrical_calculations/src/islamic.rs [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+    let midmonth = FIXED_ISLAMIC_EPOCH_FRIDAY.to_f64_date()
+        + (((year - 1) as f64) * 12.0 + month as f64 - 0.5) * MEAN_SYNODIC_MONTH;
+    let lunar_phase =
+        Astronomical::calculate_lunar_phase_at_or_before(RataDie::new(midmonth as i64));
+    Astronomical::phasis_on_or_before(RataDie::new(midmonth as i64), CAIRO, Some(lunar_phase)) + day
+        - 1
+}
+
+/// Lisp code reference: <https://github.com/EdReingold/calendar-code2/blob/1ee51ecfaae6f856b0d7de3e36e9042100b4f424/calendar.l#L6983-L6995>
+pub fn observational_islamic_from_fixed(date: RataDie) -> (i32, u8, u8) {
```
Comment (echeran) [inline utils/calendrical_calculations/src/islamic.rs:35]: naming question: why is this function named `observational_islamic...`, but the function above is `...islamic_observational`?

File: utils/calendrical_calculations/src/islamic.rs [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+        (FIXED_ISLAMIC_EPOCH_FRIDAY.to_i64_date() - 1)
+            + (year - 1) * 354
+            + div_rem_euclid_f64(3.0 + year as f64 * 11.0, 30.0).0 as i64
+            + 29 * (month - 1)
+            + div_rem_euclid_f64(month as f64, 2.0).0 as i64
+            + day,
+    )
+}
+/// Lisp code reference: <https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L2090>
+pub fn islamic_civil_from_fixed(date: RataDie) -> (i32, u8, u8) {
```
Comment (echeran) [inline utils/calendrical_calculations/src/islamic.rs:125]: more naming inconsistency: this is `islamic_civil...`, but above functions are `...saudi_islamic...`

**Review by echeran — APPROVED** [review]
LGTM modulo naming consistency nitpicks

File: components/calendar/src/hebrew.rs (hunk 2 above)
Comment (Manishearth) [reply components/calendar/src/hebrew.rs:295]: these are civil months; the names are biblical months

File: utils/calendrical_calculations/src/islamic.rs (hunk 3 above)
Comment (Manishearth) [reply utils/calendrical_calculations/src/islamic.rs:21]: a lot of this is copied boilerplate and should be cleaned up

File: utils/calendrical_calculations/src/islamic.rs (hunk 5 above)
Comment (Manishearth) [reply utils/calendrical_calculations/src/islamic.rs:125]: civil and saudi are different

Comment (Manishearth) [reply utils/calendrical_calculations/src/islamic.rs:125]: the names are from the book

File: utils/calendrical_calculations/src/islamic.rs (hunk 4 above)
Comment (Manishearth) [reply utils/calendrical_calculations/src/islamic.rs:35]: will clean up in a followup

**Comment by Manishearth** [comment]
Will do some cleanups in a followup

File: utils/calendrical_calculations/src/islamic.rs (hunk 4 above)
Comment (Manishearth) [reply utils/calendrical_calculations/src/islamic.rs:35]: ah, this is from the book. I'm going to stick to that naming

**Pull request merged**